openupgrade_scripts/scripts/purchase_stock/18.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2025 ForgeFlow S.L. (https://www.forgeflow.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from openupgradelib import openupgrade
+
+_new_columns = [
+    ("purchase.order.line", "group_id", "many2one"),
+    ("purchase.order.line", "location_final_id", "many2one"),
+    ("stock.warehouse.orderpoint", "product_supplier_id", "many2one"),
+]
+
+
+def fill_purchase_order_line_group_id(env):
```

Comment:
why the restrictions on group propagation? I'd set this unconditionally from COALESCE(po.group_id, stock_move.group_id)

Reply:
The explanation is the following:

The purchase order line only receives the group in the procurement:
![Selection_4582](https://github.com/user-attachments/assets/14a1f002-a3b7-4172-aa2b-322d4c6ec146)

But the procurement data is not saved anywhere. So, we need to go heuristic here. You cannot use freely the group of the purchase order because the group of the purchase is only the group of the procurement when the rule has 'propagate', otherwise it's the group of the stock rule.
![Selection_4583](https://github.com/user-attachments/assets/0356437e-6e46-4d7a-b654-e97a5d3a5c2b)

So, when the rule is not 'propagate', we don't have a clean way to set the group in the purchase order line.

Reply:
@hbrunn what's your opinion of my comment?

Reply:
agreed. But judging from https://github.com/odoo/odoo/commit/f9381a0611207e0fec6b957aef3b160437ecbafb and https://github.com/odoo/odoo/pull/177185/files, I think all those statements should do nothing if there's destination moves.

PS: I find links to code much more helpful than screenshots of code, makes it much more convenient to see some code in context